File: extract/tempo-and-key.py
from fileinput import filename
import json
import sys
import essentia
essentia.logLevel = 'silent'
from essentia.standard import MonoLoader, KeyExtractor, RhythmExtractor2013
import numpy as np

if len(sys.argv) < 2:
        print("Usage: python extract.py <filename>", file=sys.stderr)
        sys.exit(1)

filename = sys.argv[1]
loader = MonoLoader(filename=filename)
audio = loader()

start_time = 10
duration = 20

sample_rate = 44100
start_sample = int(start_time * sample_rate)
end_sample = start_sample + int(duration * sample_rate)

audio_segment = audio[start_sample:end_sample]

key_extractor = KeyExtractor(profileType='edmm')
key, scale, strength = key_extractor(audio_segment)

rhythm_extractor = RhythmExtractor2013()
bpm, _, _, _, _ = rhythm_extractor(audio_segment)

result = {
    "key": key,
    "bpm": round(bpm)
}
print(json.dumps(result))

File: extract/tempo-and-key.go
package extract

import (
	"encoding/json"
	"fmt"
	"os/exec"
)

func ExtractTempoAndKey(filename string) (string, int, error) {
	cmd := exec.Command("python3", "extract/tempo-and-key.py", filename)
	cmd.Stderr = nil
	output, err := cmd.Output()
	if err != nil {
		return "", 0, fmt.Errorf("error running Python script: %w - output: %s", err, string(output))
	}
	var result struct {
		Key string `json:"key"`
		BPM int    `json:"bpm"`
	}
	if err := json.Unmarshal(output, &result); err != nil {
		return "", 0, fmt.Errorf("error parsing JSON: %w", err)
	}

	return result.Key, result.BPM, nil
}
